packages/frontend/src/utils/wallet.js

```diff
@@ -273,6 +282,89 @@ class Wallet {
+        const allKeys = await this.getAccessKeys(accountId);
+        const keyInfoView = allKeys.find(({public_key}) => public_key === publicKeyString);
+
+        if (keyInfoView) {
+           if (this.isFullAccessKeyInfoView(keyInfoView)) return this.KEY_TYPES.FAK;
+           if (this.isLedgerKeyInfoView(accountId, keyInfoView)) return this.KEY_TYPES.LEDGER;
+           if (this.isMultisigKeyInfoView(accountId, keyInfoView)) return this.KEY_TYPES.MULTISIG;
+            return this.KEY_TYPES.OTHER;
+        }
+
+        throw new Error('No matching key pair for public key');
+    }
+
+    isFullAccessKeyInfoView(keyInfoView) {
+        return keyInfoView?.access_key?.permission === 'FullAccess';
+    }
+
+    isLedgerKeyInfoView(accountId, keyInfoView) {
+        const receiver_id = keyInfoView?.access_key?.permission?.FunctionCall?.receiver_id;
+        const method_names = keyInfoView?.access_key?.permission?.FunctionCall?.method_names;
+        return receiver_id === accountId && isEqual(method_names, ['__wallet__metadata']);
+    }
+
+    isMultisigKeyInfoView(accountId, keyInfoView) {
+        const receiver_id = keyInfoView?.access_key?.permission?.FunctionCall?.receiver_id;
+        const method_names = keyInfoView?.access_key?.permission?.FunctionCall?.method_names;
+        return receiver_id === accountId && isEqual(method_names, ['add_request', 'add_request_and_confirm', 'delete_request', 'confirm']);
+    }
+
+    async addExistingAccountKeyToWalletKeyStore(accountId, keyPair, ledgerHdPath) {
+        const keyType = await this.getPublicKeyType(
+            accountId,
+            keyPair.getPublicKey().toString()
+        );
+
+        switch (keyType) {
+            case this.KEY_TYPES.FAK: {
+                const keyStore = new nearApiJs.keyStores.InMemoryKeyStore();
+                await keyStore.setKey(NETWORK_ID, accountId, keyPair);
+                const newKeyPair = nearApiJs.KeyPair.fromRandom('ed25519');
+                return new nearApiJs.Account(
+                    nearApiJs.Connection.fromConfig({
+                        networkId: NETWORK_ID,
+                        provider: {
+                            type: 'JsonRpcProvider',
+                            args: { url: NODE_URL + '/' },
+                        },
+                        signer: new nearApiJs.InMemorySigner(keyStore),
+                    }),
+                    accountId
+                )
+                    .addKey(newKeyPair.getPublicKey())
+                    .then(() => this.saveAccount(accountId, newKeyPair))
+                    .then(() => {
+                        if (!this.accountId) {
+                            return this.makeAccountActive(accountId);
+                        }
+                        return this.save();
+                    });
+            }
+            case this.KEY_TYPES.MULTISIG:
+            case this.KEY_TYPES.LEDGER:
+                return this.saveAccount(accountId, keyPair)
```

Comment:
Consider `await` here as well since it's already in an `async` function and can take advantage of non-chained syntax. Just be sure to add `{}` around the `case` body so that block variables don't end up confusing scoping rules.

Reply:
Done 👍